components/layout_2020/display_list/mod.rs

```diff
@@ -1118,3 +1257,72 @@ fn create_clip_chain(
             .define_clip_chain(parent_clip_chain_id, [new_clip_id]),
     )
 }
+
+/// Compute the additional border-image area.
+pub fn resolve_border_image_outset(
+    outset: BorderImageOutset,
+    border: SideOffsets2D<f32, LayoutPixel>,
+) -> SideOffsets2D<f32, LayoutPixel> {
+    fn image_outset_for_side(outset: NonNegativeLengthOrNumber, border_width: f32) -> f32 {
+        match outset {
+            NonNegativeLengthOrNumber::Length(length) => length.px(),
+            NonNegativeLengthOrNumber::Number(factor) => border_width * factor.0,
+        }
+    }
+
+    SideOffsets2D::new(
+        image_outset_for_side(outset.0, border.top),
+        image_outset_for_side(outset.1, border.right),
+        image_outset_for_side(outset.2, border.bottom),
+        image_outset_for_side(outset.3, border.left),
+    )
+}
+
+pub fn resolve_border_image_width(
+    width: &BorderImageWidth,
+    border: SideOffsets2D<f32, LayoutPixel>,
+    border_area: Size2D<f32, LayoutPixel>,
+) -> SideOffsets2D<f32, LayoutPixel> {
+    fn image_width_for_side(
+        border_image_width: &BorderImageSideWidth,
+        border_width: f32,
+        total_length: f32,
+    ) -> f32 {
+        match border_image_width {
+            BorderImageSideWidth::LengthPercentage(v) => {
+                v.to_used_value(Au::from_f32_px(total_length)).to_f32_px()
+            },
+            BorderImageSideWidth::Number(x) => border_width * x.0,
+            BorderImageSideWidth::Auto => border_width,
+        }
+    }
+
+    SideOffsets2D::new(
+        image_width_for_side(&width.0, border.top, border_area.height),
+        image_width_for_side(&width.1, border.right, border_area.width),
+        image_width_for_side(&width.2, border.bottom, border_area.height),
+        image_width_for_side(&width.3, border.left, border_area.width),
+    )
+}
+
+/// Get the WebRender border-image slice from the style values.
+///
+/// FIXME(emilio): WR expects device pixels here... somehow?
```

Comment:
```suggestion
/// FIXME: WR expects device pixels here... somehow?
```
Is this FIXME still relevant?

Reply:
I guess not. It sort of makes sense that WebRender takes device pixels here as the values measure pixels of the raster image.